### PR TITLE
Add `translate` element

### LINKED ISSUE #1174: `translate` element
### Description

I suggest adding a `translate` function that returns content based on the current language.

`translate` accepts a [dictionary](https://typst.app/docs/reference/types/dictionary/) whoes keys are language codes and values are [content](https://typst.app/docs/reference/types/content/) translated into the corresponding language. For example:
```typst
#translate((
  en: "Theorem",
  de: "Theorem",
  fr: "Théorème",
  es: "Teorema",
))
```
returns the content corresponding to the current language.

Alternatively, `translate` accepts a [function](https://typst.app/docs/reference/types/function/) that takes the current language and region as two strings and returns content:
```typst
#translate((lang, region) => {
  if lang == "en" and region == "us" [Internationalization]
  else if lang == "en" and region == "gb" [Internationalisation]
  else if lang == "de" [Internationalisierung]
  else if lang == "fr" [Internationalisation]
  else if lang == "es" [Internacionalización]
  else { panic("Unknown language") }
})
```

### Use Case

Such a feature would be useful when creating macro-like functions.

In the following example, taken from https://typst.app/docs/reference/meta/counter/, the word "Theorem" is untranslated if the function is used in a non-English context.
```typst
#let c = counter("theorem")
#let theorem(it) = block[
  #c.step()
  *Theorem #c.display():* #it
]
```

Although it is uncommon to use multiple languages within the same document, you may want to reuse certain functions in multiple documents, which are not all written in the same language. In this case, a `translate` function would be very appreciated.

### PR DESCRIPTION
**Pull request by MDLC01**
Closes #1174

This PR introduces a new `translate` element that is rendered differently depending on the current locale (text [lang] and [region]).

As of now, there is no way to reproduce the behavior of `translate` in Typst. When get rules are added to the language, `translate` will become redundant in that its behavior will be reproducible using those. However, I believe it makes sense to have a `translate` element anyway since this is a common pattern in macros / templates, and it makes i18n easier, which is important.

## Usage

`translate` can be passed a dictionary or a function.

If passed a dictionary, each `(lang, value)` pair indicates what value to use in case the current text lang is `lang`.

```typst
#let theorem = translate((
  fr: "Théorème",
  ja: "定理",
  sv: "Teorem",
))

#set text(lang: "fr")
#theorem

#set text(lang: "ja")
#theorem

#set text(lang: "sv")
#theorem
```
Théorème // 定理 // Teorem

If passed a function, it is called every time the element is showed with the current lang and region as arguments. The element will display as the return value of the function.

```typst
#let locale = translate((lang, region) => {
  if region == none {
    lang
  } else {
    [#(lang)-#region]
  }
})

#locale

#set text(lang: "de", region: none)
#locale

#set text(lang: "fr", region: "ca")
#locale
```
en // de // fr-CA

As you may have noticed, there is no way to specify a default value in case you use the dictionary form. This is because (1) you can use the function form if you can to handle every case, and (2) it often does not make much sense to have a default value, which often ends up being the english value.

## Known issues

Due to `translate` elements being translated whenever they are displayed (which is the whole point), they may appear differently when duplicated somewhere else (typically, in an outline or a page header). This issue is similar to #1057.

[lang]: https://typst.app/docs/reference/text/text/#parameters-lang
[region]: https://typst.app/docs/reference/text/text/#parameters-region

### CODE REVIEW

File: crates/typst-library/src/text/translate.rs [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        let region = TextElem::region_in(styles);
+        if let Some(value) = self.translations().value_in(vt, lang, region)? {
+            Ok(value)
+        } else {
+            let locale = if let Some(region) = region {
+                eco_format!("{}-{}", lang.as_str(), region.as_str())
+            } else {
+                lang.as_str().into()
+            };
+            bail!(syntax::Span::detached(), "No translation available for {}", locale)
```
Comment (MDLC01) [inline crates/typst-library/src/text/translate.rs:101]: I'm not sure how to get the span corresponding to the place where the element is used. `self.span()` corresponds to where it is defined.

**Comment by laurmaedje** [comment]
While I understand the need for this, I'm not sure I like the idea of adding a stop-gap solution that we already know will be superseded / removed soon-ish. Or would you say that `translate` should stay even in presence of get rules?

**Comment by MDLC01** [comment]
I think `translate` shoud stay even in presence of get rules, as getting the current locale and returning corresponding content will likely be a common pattern. The same way we have, e.g., `NN` as a shortcut for `bb(N)` because this is a common pattern, it makes sense to have an easy way to translate content.

I fully understand, though, if you prefere to implement get rules before adding `translate`, to make sure it makes sense to have such a function at this point.

**Comment by laurmaedje** [comment]
I need to think some more about this.

Another thing:
> As you may have noticed, there is no way to specify a default value in case you use the dictionary form. This is because (1) you can use the function form if you can to handle every case, and (2) it often does not make much sense to have a default value, which often ends up being the english value.

I think providing a default is important. If `translate` is used in a package and the package doesn't provide _every_ locale ever, it will be completely incompatible with unspecified locales. Right now, a lot of Typst's built-in things are only localized for like 15 languages. Others get English. This is not ideal, but I'd argue it's better than people writing in these languages not being able to use Typst at all with the correct locale setting for text shaping.

**Comment by MDLC01** [comment]
Ok. If you think it makes sense to continue working on `translate` for now, I can add a `default` parameter.

**Comment by laurmaedje** [comment]
> Ok. If you think it makes sense to continue working on translate for now, I can add a default parameter.

I'll discuss this with @reknih and come back to you.

**Comment by reknih** [comment]
I am reluctant to ship a `translate` element before we have fully thought about the localization needs of our users. In my view, `translate` users will often be packages and the translated terms will be similar to UI strings in apps in terms of requirements. For example, imagine a package that omits figures conditionally and then prints which figures were omitted at the end of the project. This string will depend on the language not only in terms of the actual words, but also variable placement etc. Examples:

- en-US: Figure 11 was omitted.
- de-DE: 11. Figur ausgelassen. (Prefers ordinal notation, variable somewhere else)
- fr-FR: La figure 11 a été omise.
- Imagine non-arabic numerals in CJK etc.

I would like to spec a more complete system that accommodates these use cases, be it with closures for each locale or more specific tooling like format strings.

**Comment by alerque** [comment]
c.f. #216

I still recommend holding off from any half-way solutions in the way of l18n, they make it *so* much harder to do actually good localization. Do it right the first time. That means not a key-value replacement table and not an API interface with no context.

**Comment by PgBiel** [comment]
Using context proposal notation, I think this proposed element (at least, as is) would be entirely superseded by something like:

```typ
context {
  (
    en: "Hello world!",
    pt: "Olá mundo!"
  ).at(text.lang, default: "Hello world!")
}
```

**Comment by Dherse** [comment]
While I still don’t like the idea of having a `context` keyword, that does make this so easy it would not justify having this as part of the base kit. On the other hand, until @laurmaedje is done with his rework of the web app, we won’t have that syntax either way 🤷‍♂️

**Comment by MDLC01** [comment]
> Using context proposal notation, I think this proposed element (at least, as is) would be entirely superseded by something like:
> 
> ```typst
> context {
>   (
>     en: "Hello world!",
>     pt: "Olá mundo!"
>   ).at(text.lang, default: "Hello world!")
> }
> ```

Oh, I was not aware of this proposal. I agree that, if it is accepted, `translate` would become redundant.

**Comment by laurmaedje** [comment]
> Oh, I was not aware of this proposal. I agree that, if it is accepted, translate would become redundant.

It's the currently planned design for get rules, we'll have to see how it plays out. You can read more here if you're interested: https://laurmaedje.github.io/posts/types-and-context/

**Pull request closed without merging**